test/integration/operator/extension/required/required_test.go

```diff
@@ -0,0 +1,224 @@
+		extensionProvider = "ext-local"
+		dnsProvider = "dns-local"
+
+		providerExtension = &operatorv1alpha1.Extension{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: extensionPrefix + "-provider",
+				Labels: map[string]string{
+					testID: testRunID,
+				},
+			},
+			Spec: operatorv1alpha1.ExtensionSpec{
+				Resources: []gardencorev1beta1.ControllerResource{
+					{
+						Kind: "BackupBucket",
+						Type: backupBucketProvider,
+					},
+					{
+						Kind: "Extension",
+						Type: extensionProvider,
+					},
+				},
+			},
+		}
+
+		dnsExtension = &operatorv1alpha1.Extension{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: extensionPrefix + "-dns",
+				Labels: map[string]string{
+					testID: testRunID,
+				},
+			},
+			Spec: operatorv1alpha1.ExtensionSpec{
+				Resources: []gardencorev1beta1.ControllerResource{
+					{
+						Kind: "DNSRecord",
+						Type: dnsProvider,
+					},
+				},
+			},
+		}
+
+		backupBucket = &extensionsv1alpha1.BackupBucket{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: extensionPrefix + "-bucket",
+			},
+			Spec: extensionsv1alpha1.BackupBucketSpec{
+				DefaultSpec: extensionsv1alpha1.DefaultSpec{
+					Class: ptr.To[extensionsv1alpha1.ExtensionClass]("garden"),
+					Type:  backupBucketProvider,
+				},
+				Region: "region",
+				SecretRef: corev1.SecretReference{
+					Name: "test-bar-bucket",
+				},
+			},
+		}
+
+		dnsRecord = &extensionsv1alpha1.DNSRecord{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      extensionPrefix + "-record",
+				Namespace: testNamespace.Name,
+			},
+			Spec: extensionsv1alpha1.DNSRecordSpec{
+				DefaultSpec: extensionsv1alpha1.DefaultSpec{
+					Class: ptr.To[extensionsv1alpha1.ExtensionClass]("garden"),
+					Type:  dnsProvider + "-foo",
+				},
+				SecretRef: corev1.SecretReference{
+					Name: "test-foo-dns",
+				},
+				Name:       "test.example.com",
+				RecordType: extensionsv1alpha1.DNSRecordTypeA,
+				Values:     []string{"1.2.3.4"},
+			},
+		}
+
+		extension = &extensionsv1alpha1.Extension{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      extensionPrefix + "-record",
+				Namespace: testNamespace.Name,
+			},
+			Spec: extensionsv1alpha1.ExtensionSpec{
+				DefaultSpec: extensionsv1alpha1.DefaultSpec{
+					Class: ptr.To[extensionsv1alpha1.ExtensionClass]("garden"),
+					Type:  extensionProvider,
+				},
+			},
+		}
+	})
+
+	It("should reconcile the extensions and calculate the expected required status", func() {
+		By("Create extensions")
+
+		for _, ext := range []client.Object{providerExtension, dnsExtension} {
+			Expect(testClient.Create(ctx, ext)).To(Succeed())
+			log.Info("Created extension", "garden", ext.GetName())
+			DeferCleanup(func() {
+				By("Delete extension")
+				Expect(client.IgnoreNotFound(testClient.Delete(ctx, ext))).To(Succeed())
+				By("Ensure extension is gone")
+				Eventually(func() error {
+					return mgrClient.Get(ctx, client.ObjectKeyFromObject(ext), ext)
+				}).Should(BeNotFoundError())
+			})
```

Comment:
Add description to expectations in `for` loops, see https://github.com/gardener/gardener/blob/master/docs/development/testing.md#writing-tests-relevant-for-all-kinds